cmd/zrepl/mgmt_conn.c

```diff
@@ -0,0 +1,907 @@
+	}
+
+	return (rc);
+}
+
+/*
+ * One thread to serve all management connections operating in non-blocking
+ * event driven style.
+ */
+void
+uzfs_zvol_mgmt_thread(void *arg)
+{
+	char			*buf;
+	uzfs_mgmt_conn_t	*conn;
+	struct epoll_event	ev, events[MAX_EVENTS];
+	int			nfds, i, rc;
+	boolean_t		do_scan;
+
+	mutex_init(&conn_list_mtx, NULL, MUTEX_DEFAULT, NULL);
+
+	mgmt_eventfd = eventfd(0, EFD_NONBLOCK);
+	if (mgmt_eventfd < 0) {
+		perror("eventfd");
+		zk_thread_exit();
+		return;
+	}
+	epollfd = epoll_create1(0);
+	if (epollfd < 0) {
+		perror("epoll_create1");
+		zk_thread_exit();
+		return;
+	}
+	ev.events = EPOLLIN;
+	ev.data.ptr = NULL;
+	if (epoll_ctl(epollfd, EPOLL_CTL_ADD, mgmt_eventfd, &ev) == -1) {
+		perror("epoll_ctl");
+		zk_thread_exit();
+	}
+
+	prctl(PR_SET_NAME, "mgmt_conn", 0, 0, 0);
+
+	/*
+	 * The only reason to break from this loop is a failure to update FDs
+	 * in poll set. In that case we cannot guarantee consistent state.
+	 * Any other failure should be handled gracefully.
+	 */
+	while (1) {
+		do_scan = B_FALSE;
+		nfds = epoll_wait(epollfd, events, MAX_EVENTS,
+		    1000 * RECONNECT_DELAY / 2);
+		if (nfds == -1) {
+			if (errno == EINTR)
+				continue;
+			perror("epoll_wait");
+			goto exit;
+		}
+
+		for (i = 0; i < nfds; i++) {
+			conn = events[i].data.ptr;
+
+			/*
+			 * data.ptr is null only for eventfd. In that case
+			 * zinfo was created or deleted -> scan the list.
+			 */
+			if (conn == NULL) {
+				uint64_t value;
+
+				do_scan = B_TRUE;
+				/* consume the event */
+				rc = read(mgmt_eventfd, &value, sizeof (value));
+				ASSERT3P(rc, ==, sizeof (value));
+				continue;
+			}
+
+			if (events[i].events & EPOLLERR) {
+				if (conn->conn_state == CS_CONNECT) {
+					DBGCONN(conn, "Failed to connect");
+				} else {
+					fprintf(stderr, "Error on connection "
+					    "to %s:%d\n",
+					    conn->conn_host, conn->conn_port);
+				}
+				if (close_conn(conn) != 0)
+					goto exit;
+			/* tcp connected event */
+			} else if ((events[i].events & EPOLLOUT) &&
+			    conn->conn_state == CS_CONNECT) {
+				move_to_next_state(conn);
+			/* data IO */
+			} else if ((events[i].events & EPOLLIN) ||
+			    (events[i].events & EPOLLOUT)) {
+				ssize_t cnt;
+				int nbytes;
+
+				/* restore reading/writing state */
+				buf = (char *)conn->conn_buf + conn->conn_procn;
+				nbytes = conn->conn_bufsiz - conn->conn_procn;
+
+				if (events[i].events & EPOLLIN) {
+					cnt = read(conn->conn_fd, buf, nbytes);
+					DBGCONN(conn, "Read %ld bytes", cnt);
+				} else {
+					cnt = write(conn->conn_fd, buf, nbytes);
+					DBGCONN(conn, "Written %ld bytes", cnt);
+				}
+
+				if (cnt == 0) {
+					/* the other peer closed the conn */
+					if (close_conn(conn) != 0)
+						goto exit;
+				} else if (cnt < 0) {
+					if (errno == EAGAIN ||
+					    errno == EWOULDBLOCK) {
+						continue;
+					}
+					perror("read/write");
+					if (close_conn(conn) != 0)
+						goto exit;
+				} else if (cnt <= nbytes) {
```

Comment:
if cnt < nbytes, we should be going for read (or write) again..

Reply:
and that is what we do. We just increment the number of read/written bytes, leave poll events unmodified and continue polling.